implementing geom_label_aligned

suhaani-agarwal opened the pull request:
resolving #174 

This PR introduces a new geom, geom_aligned_boxes, as proposed in the issue. (name of the geom can be changed if needed)

Status: Initial implementation — still a work in progress and requires a lot of modifications and bug fixing. I am pushing the code made till now to seek guidance.
Current issue: While testing with examples, I noticed that the generated HTML renders:

``<g class="geom1_alignedboxes_plot"><g class="PANEL1"></g></g>``

The PANEL1 elements are empty, no boxes are being rendered. I suspect I may be missing a key step in correctly registering or drawing the new geom. Any guidance on what might be causing this?

tdhock commented on R/geom-aligned-boxes.R:
```diff
@@ -0,0 +1,174 @@
+)
+
+# Helper function to optimize vertical positions using quadratic programming
+optimize_vertical_positions <- function(y, heights, min_distance) {
+  n <- length(y)
+  if (n <= 1) return(y)
+  Dmat <- diag(n)
+  dvec <- y
+  Amat <- matrix(0, nrow = n * (n-1)/2, ncol = n)
+  bvec <- numeric(n * (n-1)/2)
+  k <- 1
+  for (i in 1:(n-1)) {
+    for (j in (i+1):n) {
+      Amat[k, i] <- 1
+      Amat[k, j] <- -1
+      bvec[k] <- (heights[i] + heights[j])/2 + min_distance
+      k <- k + 1
+    }
+  }
+  result <- quadprog::solve.QP(Dmat, dvec, t(Amat), bvec)
+  return(result$solution)
+}
+
+# Helper function to optimize horizontal positions using quadratic programming
+optimize_horizontal_positions <- function(x, widths, min_distance) {
+  n <- length(x)
+  if (n <= 1) return(x)
+  Dmat <- diag(n)
+  dvec <- x
```
these need to move to javascript I think

tdhock replied:
because we only know how big the text is (in pixels) after drawing it on screen.
and we need to know text size to avoid overlaps.

suhaani-agarwal commented:
I have implemented the new geom -  geom_aligned_boxes, which takes x, y, and label as required aesthetics, along with optional ones like fill, stroke, and color etc. On the JavaScript side, it dynamically measures the text size and draws a rectangle around it positioned by quadprog on renderer side, to form a label box. an example of the structure generated in the rendered HTML is as follows:
```
<g class="geom1_alignedboxes_pricesPlot">
  <g class="PANEL1">
    <g class="geom" style="opacity: 1; stroke: black; stroke-width: 2; fill: rgb(248, 118, 109);">
      <rect x="135.0775" y="226.1861" width="50.9219" height="28.5156" stroke-width="1" rx="0" ry="0"></rect>
      <text x="160.5385" y="244.4439" font-size="12px" style="text-anchor: middle;" stroke-width="1">Apple</text>
    </g>
  </g>
</g>
```

Although most of the initial implementation is complete, the positioning of these boxes using quadprog.js is not yet accurate. The elements are not restricted to the axis boundaries. I plan to refine the constraints in the quadratic programming setup to ensure correct and bounded placement. Once that is done, I will make the renderer tests for this geom.

Please let me know if the current structure and approach are appropriate or if any modifications are required.

tdhock commented on inst/htmljs/vendor/quadprog.js:
```diff
@@ -0,0 +1,600 @@
+/**
+ * JavaScript implementation of quadratic programming solver
+ * Based on the algorithm from quadprog package
+ */
```
where did this code come from?
please add a link / explanation

tdhock replied:
I understand that you copied the code from https://github.com/albertosantini/quadprog/tree/master/lib -- please add a link/reference.
"package quadprog" means a JS package, not R package, right? please make sure that is clear?

suhaani-agarwal commented:
I am still working on some edge cases for this positioning function that uses quadratic programming. (like examples with horizontal alignment are not yet being positoned in the most optimal way)

tdhock commented on inst/htmljs/vendor/quadprog.js:
```diff
@@ -598,75 +601,213 @@ function solveQP(Dmat, dvec, Amat, bvec = [], meq = 0, factorized = [0, 0]) {
+// finding the nearest possible positions to their original locations, subject to the constraint
+// that the boxes do not overlap.
+//
+// `bumpup` and `bumpleft` are heuristic fallback strategies, also inspired by directlabels.
```
i would prefer to not add bumpup and bumpleft. can you please remove? or justify their addition?

suhaani-agarwal commented:
An example of how geom_aligned_boxes look is as follows : 

https://github.com/user-attachments/assets/73a00414-eb31-420e-a98d-d58aeb12836c

 - I have to work on how the text looks like. right now it looks a little extra bold. I will correct that.
 - I may have to work on more robust boundary constraints
 - I will add this example as a test (testing that the geom aligned boxes do not collide even after interactions and some other aspects of this geom)

tdhock commented on inst/examples/WorldBank-facets-map.R:
```diff
@@ -90,7 +90,7 @@ wb.facets <- animint(
       alpha=1,
       alpha_off=0.1)+
-    geom_text(aes(
+    geom_aligned_boxes(aes(
```
I tried geom_aligned_boxes in the world bank data viz and I am very impressed at how it "just works" !!
![image](https://github.com/user-attachments/assets/9adea103-024f-41a4-bf91-bcb46bd0d1bf)
However there are a few issues.

* I was expecting white background by default (not black).
* this geom specifies hjust=1 (right justify) but we see hjust=0.5 (center justify)
* the horizontal size seems to be incorrect (Trinidad and Tobago has a lot of extra space in the box)

This is a really excellent first try though!

Can we change the name from geom_aligned_boxes to geom_label_something? for consistency with geom_label in ggplot2? maybe geom_label_aligned or geom_label_no_overlap?

suhaani-agarwal replied:
> * I was expecting white background by default (not black).

I have now changed the default fill to white.

> Can we change the name from geom_aligned_boxes to geom_label_something? for consistency with geom_label in ggplot2? maybe geom_label_aligned or geom_label_no_overlap?

I have now renamed the geom to geom_label_aligned

tdhock commented on tests/testthat/test-renderer1-geom-aligned-boxes.R:
```diff
@@ -0,0 +1,181 @@
+      height = as.numeric(attrs["height"])
+    )
+  })
+  for (i in 1:(length(box_info)-1)) {
+    for (j in (i+1):length(box_info)) {
+      box1 <- box_info[[i]]
+      box2 <- box_info[[j]]
+      x_overlap <- box1$x < (box2$x + box2$width) && (box1$x + box1$width) > box2$x
+      y_overlap <- box1$y < (box2$y + box2$height) && (box1$y + box1$height) > box2$y
```
this logic is repeated below. please write a function and use it in both cases to avoid repetitive code.

suhaani-agarwal commented on R/geom-label-aligned.R:
```diff
@@ -0,0 +1,147 @@
+#' Non-overlapping label boxes
+#'
+#' This geom creates boxes with labels that are aligned either vertically or horizontally,
+#' using quadratic programming to optimize their positions and avoid overlaps.
```
- QP is used after all data filtering and grouping, including showSelected, has been applied.
 - QP is used to optimize the positions of label boxes to avoid overlaps.
The optimization is performed along one axis only:
If alignment = "vertical", QP is used to adjust the Y positions of the labels (so they don't overlap vertically).
If alignment = "horizontal", QP is used to adjust the X positions of the labels (so they don't overlap horizontally).
 - The other axis (X for vertical, Y for horizontal) is kept fixed.
 - The alignment param can be specified by the user as "horizontal" or "vertical" while writing the geom. 
By default, the alignment is kept "vertical". 

 - I have now added information about the QP in geom-label-aligned.R in the comments.

suhaani-agarwal commented on inst/examples/geom_aligned_boxes_example.R:
```diff
@@ -0,0 +1,88 @@
+    ylab("Y")
+)
+
+animint2dir(viz3, "overlap-aligned-boxes")
```
This particular data viz is not an ideal example and I have now replaced it with the interactive data viz using world bank data that I was using for the tests. 
 - The user can specify on which axis QP will be run by the "alignment" parameter as specified here : [geom-label-aligned.R](https://github.com/animint/animint2/pull/203/files#diff-77b2d748ddd9a67b76a65f60bf619769fbccb63dbadd24cc53945bade93c85b1:~:text=%40%40%20%2D0%2C0%20%2B1%2C161%20%40%40-,%23%27%20Non%2Doverlapping%20label%20boxes,%23%27,-%23%27%20%40inheritParams%20layer) in the comments. 
 
> Is the QP run for all data points? I think there are some use cases where it would be better to run the QP several times. For example, for each unique X value, run QP on Y values of corresponding data.

Can you please clarify more on this? Are you suggesting to have a "group" parameter for the users?  so labels only affect other labels that they might realistically collide with (grouped by their Y positions when alignment="horizontal") ?

suhaani-agarwal commented:
To address the issue of non-colliding labels affecting each other's positions unnecessarily, I modified the `optimiseAlignedLabels()` function to first group labels based on their fixed coordinate (X for vertical alignment, Y for horizontal alignment) using precise bounding box calculations. This ensures labels that cannot possibly overlap are kept in separate groups and optimized independently. For horizontal alignment cases where labels share the same X position but have different Y positions (and thus don't collide), they now maintain their original positions without influencing each other, while still guaranteeing complete collision avoidance for labels that do overlap. 

This change implements your suggestion to run QP separately for groups of labels .
The solution now handles both vertical and horizontal alignment cases optimally, ensuring labels only affect others they might realistically overlap with.

Please let me know if any further modifications are required in this PR

tdhock commented:
good progress! Two comments

1. the duration option is not respected. can this be fixed please? I expected these labels to have smooth transitions, if they are specified. Can you please add a test case that fails for the current code, based on inst/examples/WorldBank-facets-map.R ? For an example of a test case involving smooth transitions, please see 
```
test-renderer2-widerect.R:406:test_that("middle of transition != after when duration=2000", {
```

2. I'm wondering if it is possible to hide the rectangular border around the text? Currently it is always shown with the same color as the text, right? do you think it is possible to do something like `geom_label_aligned(border_color=NA)` or should we make another geom, for example `geom_text_aligned()` which does not draw the border?

tdhock commented:
another thought: is it possible to implement the box as a polygon pointing to the data point? (instead of a rectangle around the text with no pointer) or is that too complicated? the code for drawing polygons with pointers in R is: 
```r
draw.polygons <- function(d,...){
  for(side in c("left", "right", "top", "bottom")){
    for(xy in c("x", "y")){
      col.name <- paste0(side, ".", xy)
      if(!col.name %in% names(d)){
        d[[col.name]] <- NA
      }
    }
  }
  if(! "box.color" %in% names(d)){
    d$box.color <- "black"
  }
  if(! "text.color" %in% names(d)){
    d$text.color <- "white"
  }
  for(i in 1:nrow(d))with(d[i,], {
    L <- list(
      x=c(left.x, left, top.x, right, right.x, right, bottom.x, left),
      y=c(left.y, top, top.y, top, right.y, bottom, bottom.y, bottom))
    for(xy.name in names(L)){
      xy <- L[[xy.name]]
      L[[xy.name]] <- xy[!is.na(xy)]
    }
    grid::grid.polygon(
      L$x, L$y,
      default.units="cm",
      gp=grid::gpar(col=box.color, fill=colour),
      name="directlabels.draw.polygon"
    )
  })
  d$colour <- d$text.color
  d
}
```

tdhock commented:
I added a new data set, data/parallelPeaks.RData, which illustrates an issue with the current code, shown below.
```r
gg_wrap <- ggplot()+
  theme_bw()+
  geom_segment(aes(
    start.seconds, process_i,
    xend=end.seconds, yend=process_i),
    data=parallelPeaks)+
  geom_point(aes(
    start.seconds, process_i),
    shape=1,
    data=parallelPeaks)+
  geom_blank(aes(
    start.seconds, process_i-1),
    shape=1,
    data=parallelPeaks)+
  geom_label_aligned(aes(
    start.seconds, process_i-0.5, label=peaks),
    alignment = "horizontal",
    data=parallelPeaks)+
  facet_wrap( ~ design + Fun, labeller=label_both, ncol=1)+
  scale_y_continuous(
    breaks=c(1,5,8,14))+
  scale_x_continuous("Time from start of computation (seconds)")
animint(gg_wrap+theme_animint(width=1500, height=1000))
```
code above has gives output below:
![image](https://github.com/user-attachments/assets/0cd0adc8-ff36-42d5-a1ce-8c80d9b08262)

* top/first panel is OK
* second and third panels are not OK. I specified geom_label_aligned(alignment="horizontal") which means, for each unique Y value , run QP on X values. but we see some groups that result from running the QP with different Y values. For example panel 2, labels 98,82,89 are in the same group, but 98,89 have the smallest Y value, and 82 has a larger Y value. can you please create a test case based on this example which fails for the current code, and then push a fix?

note that the above code has height=1000 whereas we can run the same ggplot with height=2000 as below, and we get a different result.
```r
animint(gg_wrap+theme_animint(width=1500, height=2000))
```
code above gives result below:
![image](https://github.com/user-attachments/assets/e3bfd6b5-1bf8-4ab4-b739-aff9ff487a48)
the result above is expected. the result for height=1000 should be similar (but with labels with different Y values that overlap because there is not enough vertical space). 

Does that make sense?

suhaani-agarwal commented:
> 2\. I'm wondering if it is possible to hide the rectangular border around the text? Currently it is always shown with the same color as the text, right? do you think it is possible to do something like `geom_label_aligned(border_color=NA)` or should we make another geom, for example `geom_text_aligned()` which does not draw the border?

so currently, the text is drawn on a rect, the rect border is the same color as the text and corresponds to what the user specifies in the `color` parameter for geom_label_aligned. the color inside the rect corresponds to what the user specifies in the `fill` parameter for geom_label_aligned. 

I have now also added an option of **`background_rect = FALSE/TRUE`** . By this, the user can specify if the rect in the background of each label (including border) should be drawn or not. By default, this is set as TRUE. 

An example of this is as follows : In the `worldbank-facets-map.R` , I have added **`background_rect = FALSE`** : 
<img width="1114" alt="image" src="https://github.com/user-attachments/assets/47b8dff5-b600-4b70-9348-d7aabfda9c5b" />

example of geom_label_aligned with **fill parameter:**
<img width="412" alt="image" src="https://github.com/user-attachments/assets/8eff103a-d39e-4896-ad17-9a6477d4ac0b" />

suhaani-agarwal commented:
> the result above is almost expected.
> 
> * in particular we see that 98,89 do not overlap, and 82 is above it (not involved with 98,89 position calculation), which is expected.

> the result for height=1000 should be similar (but with labels with different Y values that overlap because there is not enough vertical space).
> 
> Does that make sense?

Hi! currently, the grouping logic groups labels even if they are on **similar** y values (for horizontal alignment example) not **exactly** the same y values , so that the perpendicular overlapping can be avoided.
Currently, this "similar" y position accounts for the height of each label. If the labels are overlapping even a little in the perpendicular (y) axes, they will come under the same group and their X position will be optimised. 
 Because if we make the groups by exactly the same y values, something like below might happen :
<img width="500" height="710" alt="image" src="https://github.com/user-attachments/assets/03d137e1-35f6-47d1-b743-9986fba232c9" />
Here alignment = "horizontal"
the labels are being aligned optimally horizontally only when they are on EXACTLY the same Y value and since there are no labels lying exactly on the same y value , none are grouped together and qp is not applied on any group. 

Although, the PredictedPeaks visualization will look as expected, the overlaps will be there in the axes perpendicular to the alignment exes (here, Y) 
<img width="1000" height="895" alt="image" src="https://github.com/user-attachments/assets/3324d95a-81ff-42b9-a568-08cf4aa84403" />

Is that what is expected and should happen? Then geom_label_aligned will only be useful when we have labels on exactly the same x or y positions needing collision avoidance in only one axes.

suhaani-agarwal commented:
We can also consider adjusting the grouping logic to be less strict than it currently is. Right now, labels are grouped together if they overlap even slightly on the perpendicular axis—which is determined using the full height (for horizontal alignment) or full width (for vertical alignment) of the label. This can sometimes lead to labels being grouped and optimized together on alignment axis even when they are only slightly overlapping or close. [which is the issue right now that you pointed out in PredictedPeaks viz]

Instead, we could reduce this grouping threshold—so that labels are grouped only when they are very close in the perpendicular direction, not just slightly overlapping. This would still allow labels with nearly the same y positions (in horizontal alignment) to be grouped and optimized. 

<img width="444" height="385" alt="image" src="https://github.com/user-attachments/assets/9dc57d96-4703-4ebe-9f2a-a4697aa60b61" />

At the same time, plots like PredictedPeaks, the output would remain unchanged and as expected because here the labels are only slightly overlapping and all labels are still distinctly visible.
<img width="1007" height="893" alt="image" src="https://github.com/user-attachments/assets/4ae99d3a-dec1-4cbf-8af0-46a3d7438490" />
This adjustment helps prevent unnecessary grouping of labels that are only slightly overlapping in the perpendicular direction, while still correctly optimizing labels that are visually close and could collide.

### Which approach should be preffered : 
 - Exact-position grouping
Group labels only if they lie on the exact same position along the axis perpendicular to alignment (e.g., same y for horizontal alignment, same x for vertical). This avoids overlap along the alignment axis but may allow overlaps on the perpendicular axis even if labels are slightly misaligned.

 - Similar-position grouping (with reduced threshold)
Group labels that lie on very close positions along the perpendicular axis. The similarity threshold is smaller than the full label size, so only slightly misaligned labels are grouped. Helps avoid major perpendicular overlaps without over-grouping.

suhaani-agarwal commented:
> * we see there is no rectangle drawn around label 0, but there is a rectangle drawn around all the other labels, which is not expected. 

The issue where label = 0 was incorrectly treated as null in measureText() (resulting in a rect with zero height and width) has been fixed.
<img width="514" height="378" alt="image" src="https://github.com/user-attachments/assets/7fb8d358-fbb8-4241-a012-7a83f950112c" />

I have also added test case to verify that a rect is rendered with non-zero width and height for label = 0.

tdhock commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -0,0 +1,399 @@
+                      "Nigeria", "Mali", "South Africa", "Canada")
+
+# Filter WorldBank data
+wb <- WorldBank %>%
```
please use data.table instead of dplyr (avoid adding another import)

tdhock commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -0,0 +1,376 @@
+
+test_that("label text content is correct", {
+  box_groups <- getNodeSet(info$html, '//g[@class="geom4_labelaligned_worldbankAnim"]//g[@class="geom"]')
+  actual_texts <- sapply(box_groups, function(group) {
```
use getTextValue() in help-functions.R instead?

suhaani-agarwal commented on inst/examples/geom_label_aligned_examples.R:
```diff
@@ -0,0 +1,137 @@
+    geom_label_aligned(
+      data = label_data,
+      aes(x = fertility.rate, y = life.expectancy, label = label, fill = group, key = country),
+      alignment = "vertical", color = "#ffffd1", label_r = "9",
```
Done! I have made the input expectation for label_r numeric

suhaani-agarwal commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -198,6 +195,54 @@ test_that("labels have at least 3px vertical spacing", {
+
+info <- animint2HTML(viz_dup)
+
+test_that("labels have at least 3px vertical spacing", {
```
thanks! Since the test is intentionally failing to show what happens with the default min_distance = 0.1, should I remove it now or update it to instead test that gaps can be less than 3px (but ≥ 0.1) here?

suhaani-agarwal commented on inst/htmljs/animint.js:
```diff
@@ -1546,12 +1556,12 @@ var animint = function (to_select, json_file) {
+                    return d.optimizedPos + (d.size / 3);
                   } else {
-                    return d.scaledY - d.boxHeight * get_vjust(d) + d.boxHeight / 2 + ((d.fontsize || default_textSize) / 3);
+                    return d.scaledY - d.boxHeight * get_vjust(d) + d.boxHeight / 2 + (d.size / 3);
```
the d.size/3 adjustment helped vertically center text in labels by compensating for SVG's baseline positioning. Instead of this manual offset, I'm now using `.attr("dominant-baseline", "middle")` which handles text centering natively. This is a more reliable approach
I have removed `+ (d.size / 3)` now.
Thanks for pointing this out!

suhaani-agarwal commented:
> I added an expectation about font-size that fails using the current code on the World Bank test case.
> geom_label_aligned(size=5) has no effect (font-size: 12px is used no matter what the user specifies as size param). can you please fix?

This issue has been fixed now, the user will now be able to specify the size parameter for fontsize of the text. 
for example for geom_label_aligned(size = 5) in the same plot, we now get:

<img width="400" height="385" alt="image" src="https://github.com/user-attachments/assets/6c4416f1-0ed9-4671-bdd6-3acc08086f08" />

tdhock commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -0,0 +1,376 @@
+  # XPath to target the <text> elements inside <g class="geom" id="MediumXXX">
+  medium_label_text_xpath <- '//g[@class="geom2_labelaligned_orangeGrowth"]//g[starts-with(@id, "Medium")]//text'
+  initial_text_nodes <- getNodeSet(info$html, medium_label_text_xpath)
+  initial_font_sizes_num <- sapply(initial_text_nodes, function(node) {
+    as.numeric(gsub("px", "", xmlGetAttr(node, "font-size")))
+  })
```
can you please use getStyleValue instead? (if it works? I think it should)
```r
  ts_size <- getStyleValue(info$html, '//g[@class="geom2_labelaligned_lifeExpectancyPlot"]//text', "font-size")
```

suhaani-agarwal replied:
I’ve now updated the tests to use `getPropertyValue` because in geom_label_aligned, the font-size is directly set as an attribute on the `<text>` element, not inside the style. So getStyleValue() wouldn’t retrieve it correctly in this case.

tdhock commented on inst/htmljs/animint.js:
```diff
@@ -1429,6 +1459,117 @@ var animint = function (to_select, json_file) {
+                    if (alignment == "vertical") {
+                      return d.optimizedPos - d.boxHeight / 2;
+                    } else {
+                      return d.scaledY - d.boxHeight * get_vjust(d);
```
vjust is not supported anywhere else in the code -- please add test cases for values other than 0.5

suhaani-agarwal replied:
Done, I have added a test for vjust values 1, 0 and 0.5 for horizontal alignment.

tdhock commented on R/geom-label-aligned.R:
```diff
@@ -0,0 +1,155 @@
+#' @inheritParams geom_point
+#' @param label_r Radius of rounded corners. Defaults to 0.15 lines.
+#' @param alignment One of "vertical" (QP on Y axis) or "horizontal" (QP on X axis)
+#' @param min_distance Minimum distance between boxes in native units.
```
please clarify units for min_distance.
it is written native units, which I interpret to mean the units displayed on the axis.
However I tried size=40 with min_distance=5 and the distance between boxes is clearly smaller than 5 axis units.
<img width="354" height="555" alt="image" src="https://github.com/user-attachments/assets/7aec13a3-be99-4c99-a223-03402e09712a" />

suhaani-agarwal replied:
Thanks for pointing that out, the earlier comment was misleading. min_distance is actually in pixels, not in native/axis units. I’ve updated the documentation accordingly to reflect this.

tdhock commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -0,0 +1,376 @@
+    '//g[@class="geom4_labelaligned_worldbankAnim"]//g[@class="PANEL1"]//g[@class="geom"]//rect'
+  )[[1]]
+  attrs <- xmlAttrs(rect_node)
+  expect_equal(as.numeric(attrs[["rx"]]), 5)
+  expect_equal(as.numeric(attrs[["ry"]]), 5)
```
why is rx 5 if label_r=9 ?

suhaani-agarwal replied:
The original label_r was set to 5, which is why the rx/ry values were tested against 5 and were passing. It seems the value got updated to 9 in a later commit. I have now reverted it back to 5.

tdhock commented:
This is really excellent work! Please fix the few things I mentioned before I merge. We are almost done.

suhaani-agarwal commented on inst/htmljs/animint.js:
```diff
@@ -1429,6 +1459,117 @@ var animint = function (to_select, json_file) {
+        var fontsize = d.size;
+        var textSize = measureText(d.label, fontsize, d.angle, textStyle);
+        d.boxWidth = textSize.width;
+        d.boxHeight = textSize.height;
```
That's because getBBox() returns the bounding box of the text, which includes space below the baseline to account for descenders (like the letter 'g') and general font metrics.

For example, in the image below, Nigeria appears visually centered due to the descender in 'g', while South Africa, which lacks descenders, appears slightly higher despite having the same alignment:

<img width="243" height="106" alt="image" src="https://github.com/user-attachments/assets/e726b8f7-47ce-4961-9fc2-2342284c7787" />

We could experiment with a small manual offset for y position of `<text>` to better visually center the text--like adjusting the vertical position using + d.size / 3, which I had tried earlier. Let me know if you'd like me to explore that again

tdhock replied:
ok, that makes sense. 
it does look a bit strange for numbers which lack the descenders, but I think that we should keep this logic for now, for simplicity.

suhaani-agarwal commented:
> Also there is a problem with the smooth transitions. When the geom appears, it flies from the upper left of the plot, but it should just appear instantly (no smooth transition).
> See video:

Hi, I have updated the code to fix this issue, the labels appear instantly and dont fly in from the upper left anymore

tdhock commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -175,22 +175,17 @@ test_that("Aligned labels in timeSeries do not collide after selection/deselecti
-  expect_equal(as.numeric(attrs[["ry"]]), 5)
+  rx <- getPropertyValue(info$html, '//g[@class="geom4_labelaligned_worldbankAnim"]//rect', "rx")
+  ry <- getPropertyValue(info$html, '//g[@class="geom4_labelaligned_worldbankAnim"]//rect', "ry")
+  expect_true(all(as.numeric(rx) == 5))
```
please use expect_equal

suhaani-agarwal replied:
done!

tdhock commented:
I did some interactive testing and the new features seem to be working well.

tdhock commented:
there seems to be an issue with vjust.
typically vjust=0 means bottom alignment, and vjust=1 means top alignment.
for example consider this
```r
df <- data.frame(vjust=c(0,0.5,1))
ggplot(df)+
  geom_point(aes(vjust,vjust), color="red")+
  geom_text(aes(vjust,vjust,label=vjust,vjust=vjust), size=5)
```
<img width="249" height="137" alt="image" src="https://github.com/user-attachments/assets/85892851-e2bb-44eb-96ab-0f220838e455" />

but to get top alignment in the upper left world bank time series data viz as below
<img width="1370" height="583" alt="image" src="https://github.com/user-attachments/assets/ee3c9e27-8b2e-48b0-ab1e-6cc6de9c4236" />
I had to just `vjust=0` which is opposite of expected. 
Can you please add a test case which fails for the current code?
And then push a fix?

(by the way this vertical and horizontal label functionality is really cool!!)

suhaani-agarwal commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -177,8 +177,8 @@ test_that("Aligned labels in timeSeries do not collide after selection/deselecti
   ry <- getPropertyValue(info$html, '//g[@class="geom4_labelaligned_worldbankAnim"]//rect', "ry")
-  expect_true(all(as.numeric(rx) == 5))
-  expect_true(all(as.numeric(ry) == 5))
+  expect_equal(as.numeric(rx), rep(5, length(rx)))
```
done!

tdhock commented on tests/testthat.R:
```diff
@@ -7,31 +7,24 @@ source("helper-HTML.R")
+# Run tests
+tests_run(filter=filter)
+# Save coverage and cleanup
+if(coverage_active) {
```
I see this error in compiler tests
```
── Error ('test-compiler-ghpages.R:42:3'): animint2pages() returns list of meta-data ──
<error_stop_port_search/rlang_error/error/condition>
Error in `startup(port = port, ...)`: Chrome debugging port not open after 10 seconds.
Backtrace:
     ▆
  1. └─animint2::animint2pages(...) at test-compiler-ghpages.R:42:3
  2.   └─chromote::ChromoteSession$new()
  3.     └─chromote (local) initialize(...)
  4.       └─chromote::default_chromote_object()
  5.         ├─chromote::set_default_chromote_object(Chromote$new())
  6.         └─Chromote$new()
  7.           └─chromote (local) initialize(...)
  8.             └─Chrome$new()
  9.               └─chromote (local) initialize(...)
 10.                 └─chromote:::launch_chrome(path, args)
 11.                   └─chromote:::with_random_port(...)
 12.                     ├─base::tryCatch(...)
 13.                     │ └─base (local) tryCatchList(expr, classes, parentenv, handlers)
 14.                     │   └─base (local) tryCatchOne(expr, names, parentenv, handlers[[1L]])
 15.                     │     └─base (local) doTryCatch(return(expr), name, parentenv, handler)
 16.                     └─chromote (local) startup(port = port, ...)
 17.                       └─rlang::abort(...)
 ```
which seems to indicate that chromote is required but not available for the gh pages compiler test.
I believe tests_init() starts chromote, so here I remove the if statement, so tests_init() will be run in the compiler tests.

tdhock replied:
I'm not sure why we were not getting this error before this PR?

suhaani-agarwal replied:
Hi! actually we were getting this error before this PR and the coverage PR too, see : https://github.com/animint/animint2/actions/runs/15376944905/job/43262818512
This error automatically went away after some commits. And we were getting this error randomly in some commits like : https://github.com/animint/animint2/actions/runs/15497830316/job/43638653143 and more..
This error appears randomly sometimes and sometimes it doesnt. 

Im not sure but its maybe because the Chromote session is failing to start properly within the timeout period sometimes?

tdhock replied:
ok. so it was appearing before but randomly.
the fix I pushed seemed to have fixed it, but maybe that is random?

tdhock commented:
this is excellent work! I updated the world bank map data viz https://tdhock.github.io/2025-01-WorldBank-facets-map/ to use `geom_label_aligned`
it is really great how the text labels automatically decrease in size when there are too many of them to fit in the available space at the original size, as below.
<img width="1277" height="574" alt="image" src="https://github.com/user-attachments/assets/a5224105-c941-4040-8199-b150cefcc8ed" />

tdhock commented on tests/testthat/test-renderer1-geom-label-aligned.R:
```diff
@@ -0,0 +1,385 @@
+      clickSelects = "Tree"
+    ) +
+    scale_color_manual(
+      values = c(Fast = "#E41A1C", Medium = "#377EB8", Slow = "#4DAF4A"),
+      name = "Growth Rate"
+    ) +
+    scale_fill_manual(
+      values = c(Fast = "#E41A1C", Medium = "#377EB8", Slow = "#4DAF4A"),
+      name = "Growth Rate"
```
please refactor to an intermediate variable to avoid repeating these values

suhaani-agarwal replied:
done!

tdhock commented:
great, thanks for all your hard work on this very excellent new feature!!
